src/printer.js

```diff
@@ -3859,6 +3859,10 @@ function makeString(rawContent, enclosingQuote) {
   // Matches _any_ escape and unescaped quotes (both single and double).
   const regex = /\\([\s\S])|(['"])/g;
 
+  // Matches any unnecessarily escaped character.
+  // Adapted from https://github.com/eslint/eslint/blob/de0b4ad7bd820ade41b1f606008bea68683dc11a/lib/rules/no-useless-escape.js#L27
+  const regexUnnecessaryStringEscapes = /(^|[^\\])\\([^\\nrvtbfux\r\n\u2028\u2029"'0])/;
```

Comment:
Just noticed my solution doesn't quite work either:

```js
String.raw`\a\a`.replace(r, (match, prev, escaped) => prev + escaped) === String.raw`a\a`
```

You need [lookbehinds](https://github.com/tc39/proposal-regexp-lookbehind), but they're not in the language yet.

I fear this might not be doable with a regex, at least in a single pass.

(A hack is to just keep applying the replacement until things stop changing, but that's not ideal.)